File: src/python/sound.py
import ika
from thing import Thing

class NullSound(object):
    def __init__(self):
        self.position = 0
        self.volume = 1.0

    def Play(self):
        pass

    def Pause(self):
        pass

class RepeatableSound(object):
    def __init__(self, fname):
        self.fname = fname
        self.sounds = [ika.Sound(fname)]
        self.sounds[0].loop = False

    def Play(self):
        for s in self.sounds:
            if s.position == 0:
                s.Play()
                return

        s = ika.Sound(self.fname)
        s.loop = False
        s.Play()
        self.sounds.append(s)

# effects:

slash1, slash2, slash3 = [
    RepeatableSound('sfx/swing%i.wav' % i) for i in range(1,4)
    ]
playerHurt = NullSound()

achievement = RepeatableSound('sfx/LevelUp.wav')

menuClick = RepeatableSound('sfx/MenuClick.wav')
menuBuzz = RepeatableSound('sfx/MenuBuzz.wav')

hearthRend = RepeatableSound('sfx/HearthRend.wav')
crushingGale = RepeatableSound('sfx/CrushingGale.wav')
healingRain = RepeatableSound('sfx/HealingRain.wav')

monsterHit = RepeatableSound('sfx/MonsterHit.wav')

anklebiterStrike = ika.Sound('sfx/AnklebiterStrike.wav')
anklebiterHurt = NullSound() # ika.Sound('sfx/AnklebiterHurt.wav')
anklebiterDie = RepeatableSound('sfx/AnklebiterDie.wav')

yetiStrike = [NullSound(), NullSound()]
yetiHurt = [[ika.Sound('sfx/YetiHurt%i.wav' % i) for i in range(1,4)],
            [ika.Sound('sfx/SoulReaverHurt%i.wav' % i) for i in range(1,4)]]
yetiDie = [ika.Sound('sfx/YetiDie.wav'), ika.Sound('sfx/SoulReaverDie.wav')]

razorManeStrike = RepeatableSound('sfx/RazormaneStrike.wav')
razorManeHurt = RepeatableSound('sfx/RazormaneHurt.wav')
razorManeDie = RepeatableSound('sfx/RazormaneDie.wav')

# other effects?

class Crossfader(Thing):
    def __init__(self):
        self.oldMusic = []
        self._music = None
        self.inc = 0.01

    def _setMusic(self, value):
        assert value is not None
        self._music = value

    music = property(lambda self: self._music, _setMusic)

    def reset(self, newMusic):
        if newMusic is self.music:
            return

        if newMusic in self.oldMusic:
            self.oldMusic.remove(newMusic)

        if self.music is not None:
            if self.music not in self.oldMusic:
                self.oldMusic.append(self.music)

            self.music = newMusic
            self.music.volume = 0.0
            self.music.Play()
        else:
            self.music = newMusic
            self.music.volume = 1.0
            self.music.Play()

    def kill(self):
        if self.music:
            self.music.volume = 0.0
            self.music.Pause()
            self._music = None
            for m in self.oldMusic:
                m.volume = 0
            self.oldMusic = []

    def update(self):
        i = 0
        while i < len(self.oldMusic):
            m = self.oldMusic[i]
            m.volume -= self.inc
            if m.volume <= 0:
                m.Pause()
                self.oldMusic.pop(i)
            else:
                i += 1

        self.music.volume += self.inc

        if not self.oldMusic and self.music.volume >= 1.0:
            return True

    def draw(self):
        pass

File: src/winter/sound.ts
// TODO DO NOT COMMIT - actually use
import { Thing } from "./thing.js"

export class SoundBase {
    public position = 0
    public loop = false
    public volume = 1.0

    Play() {}
    Pause() {}
}

export class Sound extends SoundBase {
    constructor(_fileName: string) { super() }

    // TODO DO NOT COMMIT - make work
    Play() {}
    Pause() {}
}

export class NullSound extends SoundBase {
}

export class RepeatableSound extends SoundBase {
    private sounds: Sound[]

    constructor(private fileName: string) {
        super()
        this.sounds = [new Sound(fileName)]
        this.sounds[0].loop = false
    }

    Play() {
        for (let s of this.sounds) {
            if (s.position === 0) {
                s.Play()
                return
            }
        }

        const s = new Sound(this.fileName)
        s.loop = false
        s.Play()
        this.sounds.push(s)
    }
}

// effects:

export const slash1 = new RepeatableSound('sfx/swing1.wav')
export const slash2 = new RepeatableSound('sfx/swing2.wav')
export const slash3 = new RepeatableSound('sfx/swing3.wav')
export const playerHurt = new NullSound()

export const achievement = new RepeatableSound('sfx/LevelUp.wav')

export const menuClick = new RepeatableSound('sfx/MenuClick.wav')
export const menuBuzz = new RepeatableSound('sfx/MenuBuzz.wav')

export const hearthRend = new RepeatableSound('sfx/HearthRend.wav')
export const crushingGale = new RepeatableSound('sfx/CrushingGale.wav')
export const healingRain = new RepeatableSound('sfx/HealingRain.wav')

export const monsterHit = new RepeatableSound('sfx/MonsterHit.wav')

export const anklebiterStrike = new Sound('sfx/AnklebiterStrike.wav')
export const anklebiterHurt = new NullSound() // new Sound('sfx/AnklebiterHurt.wav')
export const anklebiterDie = new RepeatableSound('sfx/AnklebiterDie.wav')

export const yetiStrike = new NullSound()
export const yetiHurt = [
    new Sound('sfx/YetiHurt1.wav'),
    new Sound('sfx/YetiHurt2.wav'),
    new Sound('sfx/YetiHurt3.wav'),
]
export const yetiDie = new Sound('sfx/YetiDie.wav')

export const soulReaverStrike = new NullSound()
export const soulReaverHurt = [
    new Sound('sfx/SoulReaverHurt1.wav'),
    new Sound('sfx/SoulReaverHurt2.wav'),
    new Sound('sfx/SoulReaverHurt3.wav'),
]
export const soulReaverDie = new Sound('sfx/SoulReaverDie.wav')

export const razorManeStrike = new RepeatableSound('sfx/RazormaneStrike.wav')
export const razorManeHurt = new RepeatableSound('sfx/RazormaneHurt.wav')
export const razorManeDie = new RepeatableSound('sfx/RazormaneDie.wav')

// other effects?

export class Crossfader extends Thing {
    private oldMusic: Sound[] = []
    public music: Sound | null = null
    private inc = 0.01
    constructor() { super() }

    reset(newMusic: Sound | null) {
        if (newMusic === this.music) {
            return
        }

        if (newMusic !== null && this.oldMusic.includes(newMusic)) {
            this.oldMusic = this.oldMusic.filter(m => m !== newMusic)
        }

        if (this.music !== null) {
            if (!this.oldMusic.includes(this.music)) {
                this.oldMusic.push(this.music)
            }

            this.music = newMusic
            if (this.music !== null) {
                this.music.volume = 0.0
                this.music.Play()
            }
        } else {
            this.music = newMusic
            if (this.music !== null) {
                this.music.volume = 1.0
                this.music.Play()
            }
        }
    }

    kill() {
        if (this.music) {
            this.music.volume = 0.0
            this.music.Pause()
            this.music = null
            for (let m of this.oldMusic) {
                m.volume = 0
            }
            this.oldMusic = []
        }
    }

    update() {
        const newOldMusic: Sound[] = []
        for (let m of this.oldMusic) {
            m.volume -= this.inc
            if (m.volume <= 0) {
                m.Pause()
            } else {
                newOldMusic.push(m)
            }
        }
        this.oldMusic = newOldMusic
        if (this.music !== null) {
            this.music.volume = Math.min(1.0, this.music.volume + this.inc)
        }

        return (this.oldMusic.length === 0 && (this.music === null || this.music.volume >= 1.0))
    }
}
